File: rust/src/graph.rs
// Copyright 2024 Bewusstsein Labs

#![warn(private_bounds)]

pub mod traverser;

//: Standard
use std::{
    cmp::{ Eq, Ord, PartialEq },
    collections::{ BTreeMap, BTreeSet, VecDeque },
    marker::PhantomData,
    ops::{ Deref, DerefMut, Not }
};

use thiserror::Error;

#[derive(Error, Debug)]
pub enum Error {
    #[error("Graph Error: Node Not Found")]
    NodeNotFound,
    #[error("Graph Error: Node Already Exists")]
    NodeAlreadyExists,
    #[error("Graph Error: Edge Not Found")]
    EdgeNotFound,
    #[error("Graph Error: Edge Already Exists")]
    EdgeAlreadyExists,
}

pub type AdjacencyData<I, E> = BTreeMap<I, E>;

#[derive( Debug, Clone, PartialEq, Eq )]
pub struct NodeData<I, N, E> {
    data: N,
    adjacencies: AdjacencyData<I, E>
}

impl<I, N, E> NodeData<I, N, E> {
    pub fn new( data: N ) -> Self {
        Self {
            data,
            adjacencies: BTreeMap::new()
        }
    }

    #[inline]
    pub fn data( &self ) -> &N {
        &self.data
    }

    #[inline]
    pub fn data_mut( &mut self ) -> &mut N {
        &mut self.data
    }

    #[inline]
    pub fn adjacencies( &self ) -> &AdjacencyData<I, E> {
        &self.adjacencies
    }

    #[inline]
    pub fn adjacencies_mut( &mut self ) -> &mut AdjacencyData<I, E> {
        &mut self.adjacencies
    }
}

#[derive( Debug, Clone, PartialEq, Eq )]
pub(crate) struct GraphData<I, N, E>( BTreeMap<I, NodeData<I, N, E>> );

impl<I, N, E> Deref for GraphData<I, N, E> {
    type Target = BTreeMap<I, NodeData<I, N, E>>;
    fn deref( &self ) -> &Self::Target {
        &self.0
    }
}

impl<I, N, E> DerefMut for GraphData<I, N, E> {
    fn deref_mut( &mut self ) -> &mut Self::Target {
        &mut self.0
    }
}

impl<I, N, E> GraphData<I, N, E>
where
    I: Clone + Ord,
    Self: Deref<Target=BTreeMap<I, NodeData<I, N, E>>> + DerefMut<Target=BTreeMap<I, NodeData<I, N, E>>>
{
    pub fn add_node( &mut self, id: I, data: N ) -> Result<(), Error> {
        self.contains_key( &id )
            .then_some( Err( Error::NodeAlreadyExists ) )
            .unwrap_or_else( || { self.insert( id, NodeData::new( data ) ); Ok( () ) } )
    }

    pub fn get_node( &self, id: I ) -> Option<&N> {
        self.get( &id ).map( |node| node.data() )
    }

    pub fn get_node_mut( &mut self, id: I ) -> Option<&mut N> {
        self.get_mut( &id ).map( |node| node.data_mut() )
    }

    pub fn get_nodes( &self ) -> &BTreeMap<I, NodeData<I, N, E>> {
        self
    }

    pub fn get_nodes_mut( &mut self ) -> &mut BTreeMap<I, NodeData<I, N, E>> {
        self
    }

    pub fn contains_node( &self, id: I ) -> bool {
        self.contains_key( &id )
    }

    pub fn remove_node( &mut self, id: I ) -> Result<NodeData<I, N, E>, Error> {
        self.remove( &id ).ok_or( Error::NodeNotFound )
    }

    pub fn delete_node( &mut self, id: I ) -> Result<(), Error> {
        self.remove( &id ).ok_or( Error::NodeNotFound ).map( |_| () )
    }

    pub fn add_edge(&mut self, id1: I, id2: I, data: E) -> Result<(), Error> {
        self.get_mut( &id1 ).ok_or( Error::NodeNotFound )
            .and_then( |node1| node1.adjacencies().contains_key( &id2 )
                .then_some( Err( Error::EdgeAlreadyExists ) )
                .unwrap_or_else( || { node1.adjacencies_mut().insert( id2, data ); Ok( () ) } )
            )
    }

    pub fn get_edge( &self, id1: I, id2: I ) -> Option<&E> {
        self.get( &id1 ).and_then( |node| node.adjacencies().get( &id2 ) )
    }

    pub fn get_edge_mut( &mut self, id1: I, id2: I ) -> Option<&mut E> {
        self.get_mut( &id1 ).and_then( |node| node.adjacencies_mut().get_mut( &id2 ) )
    }

    pub fn get_edges( &self, id: I ) -> Option<&AdjacencyData<I, E>> {
        self.get( &id ).map( |node| node.adjacencies() )
    }

    pub fn get_edges_mut( &mut self, id: I ) -> Option<&mut AdjacencyData<I, E>> {
        self.get_mut( &id ).map( |node| node.adjacencies_mut() )
    }

    pub fn contains_edge( &self, id1: I, id2: I ) -> bool {
        self.get( &id1 ).map_or( false, |node| node.adjacencies().contains_key( &id2 ) )
    }

    pub fn remove_edge( &mut self, id1: I, id2: I ) -> Result<E, Error> {
        self.get_mut( &id1 ).ok_or( Error::NodeNotFound )
            .and_then( |node| node.adjacencies_mut().remove( &id2 ).ok_or( Error::EdgeNotFound ) )
    }

    pub fn delete_edge( &mut self, id1: I, id2: I ) -> Result<(), Error> {
        self.get_mut( &id1 ).ok_or( Error::NodeNotFound )
            .and_then( |node| node.adjacencies_mut().remove( &id2 )
                .ok_or( Error::EdgeNotFound ).map( |_| () )
            )
    }

    pub fn clear_edges( &mut self ) {
        self.values_mut().for_each( |node| node.adjacencies_mut().clear() );
    }
}

pub(crate) trait GraphAccess<'a, I, N, E>
{
    fn data( &'a self ) -> &'a GraphData<I, N, E>;
    fn data_mut( &'a mut self ) -> &'a mut GraphData<I, N, E>;
}

pub trait GraphTraits<'a, I, N, E>
where
    I: 'a + Clone + Ord,
    N: 'a + PartialEq,
    E: 'a + PartialEq,
    Self: GraphAccess<'a, I, N, E>
{
    fn add_node( &'a mut self, id: I, data: N ) -> Result<(), Error> {
        self.data_mut().add_node( id, data )
    }

    fn get_node( &'a self, id: I ) -> Option<&'a N> {
        self.data().get_node( id )
    }

    fn get_node_mut( &'a mut self, id: I ) -> Option<&'a mut N> {
        self.data_mut().get_node_mut( id )
    }

    fn nodes( &'a self ) -> &'a BTreeMap<I, NodeData<I, N, E>> {
        self.data().get_nodes()
    }

    fn nodes_mut( &'a mut self ) -> &'a mut BTreeMap<I, NodeData<I, N, E>> {
        self.data_mut().get_nodes_mut()
    }

    fn contains_node( &'a self, id: I ) -> bool {
        self.data().contains_node( id )
    }

    fn remove_node( &'a mut self, id: I ) -> Result<NodeData<I, N, E>, Error> {
        self.data_mut().remove_node( id )
    }

    fn delete_node( &'a mut self, id: I ) -> Result<(), Error> {
        self.data_mut().delete_node( id )
    }

    fn add_edge(&'a mut self, node1: I, node2: I, data: E) -> Result<(), Error> {
        self.data_mut().add_edge( node1, node2, data )
    }

    fn get_edge( &'a self, id1: I, id2: I ) -> Option<&'a E> {
        self.data().get_edge( id1, id2 )
    }

    fn get_edge_mut( &'a mut self, id1: I, id2: I ) -> Option<&'a mut E> {
        self.data_mut().get_edge_mut( id1, id2 )
    }

    fn edges( &'a self, id: I ) -> Option<&'a AdjacencyData<I, E>> {
        self.data().get_edges( id )
    }

    fn edges_mut( &'a mut self, id: I ) -> Option<&'a mut AdjacencyData<I, E>> {
        self.data_mut().get_edges_mut( id )
    }

    fn contains_edge( &'a self, id1: I, id2: I ) -> bool {
        self.data().contains_edge( id1, id2 )
    }

    fn remove_edge( &'a mut self, id1: I, id2: I ) -> Result<E, Error> {
        self.data_mut().remove_edge( id1, id2 )
    }

    fn delete_edge( &'a mut self, node1: I, node2: I ) -> Result<(), Error> {
        self.data_mut().delete_edge( node1, node2 )
    }

    fn clear( &'a mut self ) {
        self.data_mut().clear();
    }

    fn clear_edges( &'a mut self ) {
        self.data_mut().clear_edges();
    }

    fn is_complete( &'a self ) -> bool {
        for ( node, neighbors ) in self.data().iter() {
            if neighbors.adjacencies().len() != self.data().len() - 1 {
                return false;
            }
            for neighbor in neighbors.adjacencies().keys() {
                if !self.data().get( neighbor ).map_or( false, |n| n.adjacencies().contains_key( node ) ) {
                    return false;
                }
            }
        }
        true
    }

    fn is_empty( &'a self ) -> bool {
        self.data().values().all( |neighbors| neighbors.adjacencies().is_empty() )
    }

    fn is_trivial( &'a self ) -> bool {
        self.data().len() == 1 && self.data().values().next().map_or( false, |neighbors| neighbors.adjacencies().is_empty() )
    }

    fn is_null( &'a self ) -> bool {
        self.data().is_empty()
    }

    fn is_child_node( &'a self, node_1: I ) -> bool {
        self.data().contains_key( &node_1 )
    }

    fn is_subgraph(&'a self, subgraph: &'a Self) -> bool {
        subgraph.data().iter().all( |(node, neighbors)| {
            self.data().get( node ).map_or( false, |graph_node| {
                neighbors.adjacencies().keys().all( |key| graph_node.adjacencies().contains_key( key ) )
            })
        })
    }

    fn is_proper_subgraph( &'a self, subgraph: &'a Self ) -> bool {
        self.data() != subgraph.data() && self.is_subgraph( subgraph )
    }

    fn is_improper_subgraph( &'a self, subgraph: &'a Self ) -> bool {
        self.data() == subgraph.data()
    }

    fn is_spanning_subgraph( &'a self, subgraph: &'a Self ) -> bool {
        self.data().len() == subgraph.data().len() && self.is_subgraph( subgraph )
    }

    fn are_adjacent_nodes( &'a self, node_1: I, node_2: I ) -> bool {
        self.is_child_node( node_1.clone() )
            && self.is_child_node( node_2.clone() )
            && self.data().get( &node_1 ).unwrap().adjacencies().contains_key( &node_2 )
    }

    fn are_adjacent_edges( &'a self, node_1: I, node_2: I, node_3: I ) -> bool {
        self.are_adjacent_nodes( node_1, node_2.clone() )
            && self.are_adjacent_nodes( node_2, node_3 )
    }

    fn order( &'a self ) -> usize {
        self.data().len()
    }

    fn size( &'a self ) -> usize {
        self.data().values().map( |neighbors| neighbors.adjacencies().len() ).sum::<usize>() / 2
    }
}

pub trait GraphType {}

#[derive( Debug, Clone, PartialEq, Eq )]
pub struct Graph<T, I, N, E>
where
    T: GraphType
{
    data: GraphData<I, N, E>,
    t: PhantomData<T>
}

impl<T, I, N, E> Graph<T, I, N, E>
where
    T: GraphType
{
    pub fn new() -> Self {
        Self {
            data: GraphData( BTreeMap::new() ),
            t: PhantomData
        }
    }
}

impl<T, I, N, E> Default for Graph<T, I, N, E>
where
    T: GraphType
{
    fn default() -> Self {
        Self::new()
    }
}

impl<'a, T, I, N, E> GraphAccess<'a, I, N, E> for Graph<T, I, N, E>
where
    T: GraphType
{
    #[inline]
    fn data( &'a self ) -> &'a GraphData<I, N, E> {
        &self.data
    }

    #[inline]
    fn data_mut( &'a mut self ) -> &'a mut GraphData<I, N, E> {
        &mut self.data
    }
}

File: cpp/impl/graph.cpp
// Copyright 2024 Shane W. Mulcahy

#include <iostream>
#include <vector>
#include <set>
#include <map>
#include <fstream>

#include "intf/graph.hpp"

namespace atom {

    namespace graph {

        std::ostream& operator<<( std::ostream& os, const Error& error ) {
            switch ( error ) {
                case Error::NONE:
                    os << "No error";
                    break;
                case Error::NODE_NOT_FOUND:
                    os << "Node not found";
                    break;
                case Error::EDGE_NOT_FOUND:
                    os << "Edge not found";
                    break;
                case Error::NODE_ALREADY_EXISTS:
                    os << "Node already exists";
                    break;
                case Error::EDGE_ALREADY_EXISTS:
                    os << "Edge already exists";
                    break;
                case Error::NODES_NOT_ADJACENT:
                    os << "Nodes not adjacent";
                    break;
                case Error::NODES_ALREADY_ADJACENT:
                    os << "Nodes already adjacent";
                    break;
                default:
                    os << "Unknown error";
                    break;
            }
            return os;
        }

        template<typename T>
        void bfs( const Graph<T>& graph, const T start_node ) {
            std::set<T> visited;
            std::vector<T> queue;
            queue.push_back( start_node );
            visited.insert( start_node );
            while ( !queue.empty() ) {
                T node = queue.front();
                queue.erase( queue.begin() );
                std::cout << node << std::endl;
                for ( const T& neighbor : graph.adjacencies.at( node ) ) {
                    if ( visited.find( neighbor ) == visited.end() ) {
                        queue.push_back( neighbor );
                        visited.insert( neighbor );
                    }
                }
            }
        }

        template<typename T>
        void dfs( const Graph<T>& graph, const T start_node ) {
            std::set<T> visited;
            std::vector<T> stack;
            stack.push_back( start_node );
            visited.insert( start_node );
            while ( !stack.empty() ) {
                T node = stack.back();
                stack.pop_back();
                std::cout << node << std::endl;
                for ( const T& neighbor : graph.adjacencies.at( node ) ) {
                    if ( visited.find( neighbor ) == visited.end() ) {
                        stack.push_back( neighbor );
                        visited.insert( neighbor );
                    }
                }
            }
        }

        template<typename T>
        Error Graph<T>::add_node( T node ) {
            if ( this->adjacencies.find( node ) != this->adjacencies.end() ) {
                return Error::NODE_ALREADY_EXISTS;
            }
            this->adjacencies.insert( std::make_pair( node, std::set<T>() ) );
            return Error::NONE;
        }

        template<typename T>
        Error Graph<T>::remove_node( T node ) {
            if ( this->adjacencies.find( node ) == this->adjacencies.end() ) {
                return Error::NODE_NOT_FOUND;
            }
            this->adjacencies.erase( node );
            return Error::NONE;
        }

        template<typename T>
        Error Graph<T>::add_edge( T node_1, T node_2 ) {
            if ( this->adjacencies.find( node_1 ) == this->adjacencies.end() ) {
                return Error::NODE_NOT_FOUND;
            }
            if ( this->adjacencies.find( node_2 ) == this->adjacencies.end() ) {
                return Error::NODE_NOT_FOUND;
            }
            if ( this->adjacencies.at( node_1 ).find( node_2 ) != this->adjacencies.at( node_1 ).end() ) {
                return Error::NODES_ALREADY_ADJACENT;
            }
            this->adjacencies.at( node_1 ).insert( node_2 );
            this->adjacencies.at( node_2 ).insert( node_1 );
            return Error::NONE;
        }

        template<typename T>
        Error Graph<T>::remove_edge( T node_1, T node_2 ) {
            if ( this->adjacencies.find( node_1 ) == this->adjacencies.end() ) {
                return Error::NODE_NOT_FOUND;
            }
            if ( this->adjacencies.find( node_2 ) == this->adjacencies.end() ) {
                return Error::NODE_NOT_FOUND;
            }
            if ( this->adjacencies.at( node_1 ).find( node_2 ) == this->adjacencies.at( node_1 ).end() ) {
                return Error::NODES_NOT_ADJACENT;
            }
            this->adjacencies.at( node_1 ).erase( node_2 );
            this->adjacencies.at( node_2 ).erase( node_1 );
            return Error::NONE;
        }

        template<typename T>
        void Graph<T>::clear() {
            this->adjacencies.clear();
        }

        template<typename T>
        void Graph<T>::clear_edges() {
            for ( auto& adjacency : this->adjacencies ) {
                adjacency.second.clear();
            }
        }

        template<typename T>
        std::ostream& operator<<( std::ostream& os, const Graph<T>& graph ) {
            for ( const auto& adjacency : graph.adjacencies ) {
                os << "Node " << adjacency.first << " is connected to: " << std::endl;
                for ( const auto& node : adjacency.second ) {
                    os << "\t- Node " << node << std::endl;
                }
            }
            return os;
        }

        template<typename T> bool operator==( const Graph<T>& lhs, const Graph<T>& rhs ) {
            return lhs.adjacencies == rhs.adjacencies;
        }

        template<typename T> bool operator!=( const Graph<T>& lhs, const Graph<T>& rhs ) {
            return lhs.adjacencies != rhs.adjacencies;
        }

        template<typename T> bool operator<=( const Graph<T>& lhs, const Graph<T>& rhs ) {
            return is_subgraph( rhs, lhs );
        }

        template<typename T> bool operator>=( const Graph<T>& lhs, const Graph<T>& rhs ) {
            return is_subgraph( lhs, rhs );
        }

        template<typename T> bool operator<( const Graph<T>& lhs, const Graph<T>& rhs ) {
            return is_proper_subgraph( rhs, lhs );
        }

        template<typename T> bool operator>( const Graph<T>& lhs, const Graph<T>& rhs ) {
            return is_proper_subgraph( lhs, rhs );
        }

        template<typename T>
        Graph<T> operator+( const Graph<T>& lhs, const Graph<T>& rhs ) {
            Graph<T> result;
            for ( const auto& [ node, neighbors ] : lhs.adjacencies ) {
                result.adjacencies[ node ] = neighbors;
            }
            for ( const auto& [ node, neighbors ] : rhs.adjacencies ) {
                if ( result.adjacencies.find( node ) == result.adjacencies.end() ) {
                    result.adjacencies[ node ] = neighbors;
                } else {
                    result.adjacencies[ node ].insert( neighbors.begin(), neighbors.end() );
                }
            }
            return result;
        }

        template<typename T>
        Graph<T> operator-( const Graph<T>& lhs, const Graph<T>& rhs ) {
            Graph<T> result = lhs;
            for ( const auto& [ node, neighbors ] : rhs.adjacencies ) {
                if ( result.adjacencies.find( node ) != result.adjacencies.end() ) {
                    for ( const auto& neighbor : neighbors ) {
                        result.adjacencies.at( node ).erase( neighbor );
                        if ( result.adjacencies.at( neighbor ).find( node ) != result.adjacencies.at( neighbor ).end() ) {
                            result.adjacencies.at( neighbor ).erase( node );
                        }
                    }
                    if ( result.adjacencies.at( node ).empty() ) {
                        result.adjacencies.erase( node );
                    }
                }
            }
            return result;
        }

        template<typename T>
        bool is_complete( const Graph<T>& graph ) {
            size_t n = graph.adjacencies.size();
            for ( const auto& [ node, neighbors ] : graph.adjacencies ) {
                if ( neighbors.size() != n - 1 ) {
                    return false;
                }
                for ( const T& neighbor : neighbors ) {
                    // The neighbor should also have the current node in its adjacency list
                    if (graph.adjacencies.at( neighbor ).find( node ) == graph.adjacencies.at( neighbor ).end() ) {
                        return false;
                    }
                }
            }
            return true;
        }

        template<typename T>
        bool is_empty( const Graph<T>& graph ) {
            if ( graph.adjacencies.size() == 0 ) {
                return true;
            }
            else {
                for ( const auto& [ node, neighbors ] : graph.adjacencies ) {
                    if ( neighbors.size() != 0 ) {
                        return false;
                    }
                }
            }
            return true;
        }

        template<typename T>
        bool is_trivial( const Graph<T>& graph ) {
            if ( graph.adjacencies.size() == 1 ) {
                for ( const auto& [ node, neighbors ] : graph.adjacencies ) {
                    if ( neighbors.size() == 0 ) {
                        return true;
                    }
                }
            }
            return false;
        }

        template<typename T>
        bool is_null( const Graph<T>& graph ) {
            if ( graph.adjacencies.size() == 0 ) {
                return true;
            }
            return false;
        }

        template<typename T>
        bool is_child_node( const Graph<T>& graph, const T node_1 ) {
            if ( graph.adjacencies.find( node_1 ) == graph.adjacencies.end() ) {
                return false;
            }
            return true;
        }

        template<typename T>
        bool is_subgraph( const Graph<T>& graph, const Graph<T>& subgraph ) {
            for ( const auto& [ node, neighbors ] : subgraph.adjacencies ) {
                if ( graph.adjacencies.find( node ) == graph.adjacencies.end() ) {
                    return false;
                }
                else if ( !std::includes( graph.adjacencies.at( node ).begin(), graph.adjacencies.at( node ).end(), neighbors.begin(), neighbors.end() ) ) {
                    return false;
                }
            }
            return true;
        }

        template<typename T>
        bool is_proper_subgraph( const Graph<T>& graph, const Graph<T>& subgraph ) {
            if ( graph == subgraph ) {
                return false;
            }
            else if ( is_subgraph( graph, subgraph ) ) {
                return true;
            }
            return false;
        }

        template<typename T>
        bool is_improper_subgraph( const Graph<T>& graph, const Graph<T>& subgraph ) {
            return graph.adjacencies == subgraph.adjacencies;
        }

        template<typename T>
        bool is_spanning_subgraph( const Graph<T>& graph, const Graph<T>& subgraph ) {
            if ( graph.adjacencies.size() != subgraph.adjacencies.size() ) {
                return false;
            }
            else if ( is_subgraph( graph, subgraph ) ) {
                return true;
            }
            return false;
        }

        template<typename T>
        bool are_adjacent_nodes( const Graph<T>& graph, const T node_1, const T node_2 ) {
            if ( is_child_node( graph, node_1 ) ) {
                return false;
            }
            if ( is_child_node( graph, node_2 ) ) {
                return false;
            }
            if ( graph.adjacencies.at( node_1 ).find( node_2 ) == graph.adjacencies.at( node_1 ).end() ) {
                return false;
            }
            return true;
        }

        template<typename T>
        bool are_adjacent_edges( const Graph<T>& graph, const T node_1, const T node_2, const T node_3 ) {
            if ( are_adjacent_nodes( graph, node_1, node_2 ) ) {
                return false;
            }
            if ( are_adjacent_nodes( graph, node_2, node_3 ) ) {
                return false;
            }
            return true;
        }

        template<typename T>
        size_t order( const Graph<T>& graph ) {
            return graph.adjacencies.size();
        }

        template<typename T>
        size_t size( const Graph<T>& graph ) {
            size_t size = 0;
            for ( const auto& [ node, neighbors ] : graph.adjacencies ) {
                size += neighbors.size();
            }
            return size / 2;
        }

        template<typename T>
        void generate_dot_graph( const Graph<T>& graph, const std::string& filename ) {
            std::ofstream file( filename );
            file << "graph {" << std::endl;
            for ( const auto& [ node, neighbors ] : graph.adjacencies ) {
                for ( const T& neighbor : neighbors ) {
                    file << "\t" << node << " -- " << neighbor << ";" << std::endl;
                }
            }
            file << "}" << std::endl;
        }

        template class Graph<Id>;
        template std::ostream&  operator<<( std::ostream& os, const Graph<Id>& graph );
        template bool           operator==( const Graph<Id>& lhs, const Graph<Id>& rhs );
        template bool           operator!=( const Graph<Id>& lhs, const Graph<Id>& rhs );
        template bool           operator<=( const Graph<Id>& lhs, const Graph<Id>& rhs );
        template bool           operator>=( const Graph<Id>& lhs, const Graph<Id>& rhs );
        template bool           operator<( const Graph<Id>& lhs, const Graph<Id>& rhs );
        template bool           operator>( const Graph<Id>& lhs, const Graph<Id>& rhs );
        template Graph<Id>      operator+( const Graph<Id>& lhs, const Graph<Id>& rhs );
        template Graph<Id>      operator-( const Graph<Id>& lhs, const Graph<Id>& rhs );
        template bool           is_complete( const Graph<Id>& graph );
        template bool           is_empty( const Graph<Id>& graph );
        template bool           is_trivial( const Graph<Id>& graph );
        template bool           is_null( const Graph<Id>& graph );
        template bool           is_child_node( const Graph<Id>& graph, const Id node_1 );
        template bool           is_subgraph( const Graph<Id>& lhs, const Graph<Id>& rhs );
        template bool           is_proper_subgraph( const Graph<Id>& lhs, const Graph<Id>& rhs );
        template bool           is_improper_subgraph( const Graph<Id>& lhs, const Graph<Id>& rhs );
        template bool           is_spanning_subgraph( const Graph<Id>& lhs, const Graph<Id>& rhs );
        template bool           are_adjacent_nodes( const Graph<Id>& graph, const Id node_1, const Id node_2 );
        template bool           are_adjacent_edges( const Graph<Id>& graph, const Id node_1, const Id node_2, const Id node_3 );
        template size_t         order( const Graph<Id>& graph );
        template size_t         size( const Graph<Id>& graph );
        template void           generate_dot_graph( const Graph<Id>& graph, const std::string& filename );

        template class Graph<std::string>;
        template std::ostream&      operator<<( std::ostream& os, const Graph<std::string>& graph );
        template bool               operator==( const Graph<std::string>& lhs, const Graph<std::string>& rhs );
        template bool               operator!=( const Graph<std::string>& lhs, const Graph<std::string>& rhs );
        template bool               operator<=( const Graph<std::string>& lhs, const Graph<std::string>& rhs );
        template bool               operator>=( const Graph<std::string>& lhs, const Graph<std::string>& rhs );
        template bool               operator<( const Graph<std::string>& lhs, const Graph<std::string>& rhs );
        template bool               operator>( const Graph<std::string>& lhs, const Graph<std::string>& rhs );
        template Graph<std::string> operator+( const Graph<std::string>& lhs, const Graph<std::string>& rhs );
        template Graph<std::string> operator-( const Graph<std::string>& lhs, const Graph<std::string>& rhs );
        template bool               is_complete( const Graph<std::string>& graph );
        template bool               is_empty( const Graph<std::string>& graph );
        template bool               is_trivial( const Graph<std::string>& graph );
        template bool               is_null( const Graph<std::string>& graph );
        template bool               is_child_node( const Graph<std::string>& graph, std::string const node_1 );
        template bool               is_subgraph( const Graph<std::string>& lhs, const Graph<std::string>& rhs );
        template bool               is_proper_subgraph( const Graph<std::string>& lhs, const Graph<std::string>& rhs );
        template bool               is_improper_subgraph( const Graph<std::string>& lhs, const Graph<std::string>& rhs );
        template bool               is_spanning_subgraph( const Graph<std::string>& lhs, const Graph<std::string>& rhs );
        template bool               are_adjacent_nodes( const Graph<std::string>& graph, const std::string node_1, const std::string node_2 );
        template bool               are_adjacent_edges( const Graph<std::string>& graph, const std::string node_1, const std::string node_2, const std::string node_3 );
        template size_t             order( const Graph<std::string>& graph );
        template size_t             size( const Graph<std::string>& graph );
        template void               generate_dot_graph( const Graph<std::string>& graph, const std::string& filename );

    } // namespace graph

} // namespace atom